### PR TITLE
Generate, upload, and display thumbnail of shots.

### LINKED ISSUE #3282: Create thumbnail of shot
We store just the full-resolution shot, which means that My Shots shows really large images as thumbnails. We should generate a thumbnail on shot generation.

This requires:

- [ ] Generating a thumbnail version of the shot
- [ ] Adding a thumbnail parameter to the shot JSON
- [ ] Making sure that thumbnail is moved to S3
- [ ] Making sure the thumbnail is managed like other S3 images (in the images table, etc)

Probably some work on clipRewrites is in order, as it's not very good at finding images in multiple locations in the shot object.

### PR DESCRIPTION
**Pull request by chenba**
- The thumbnails are generated on the browser.
- The thumbnails are scaled and cropped in the same fashion as how My Shots display shots as thumbnails currently; they are not simply scaled down copies of the shot image.
- The thumbnails are always PNGs.
- This patch piggyback on some existing thumbnail related code (that did not seem to be in use?).  That code inserts the thumbnails into the `images` table and allows the thumbnails to be managed identically to the shot images.
- My Shot will fall back to using the full size shot image if a thumbnail isn't available.

Fix #3282

### CODE REVIEW

**Comment by chenba** [comment]
In the patch for #3728 (PR #3819), I made some changes on how the shots are displayed as thumbnails.  I will also simplify how this patch scale and crop the shots to generate thumbnails; it's a bit overwrought at the moment.

`[WIP]` will be in the title of this PR until that is ready.

File: addon/webextension/background/takeshot.js [hunk 1]
```diff
@@ -33,8 +34,9 @@ this.takeshot = (function() {
     }
     let convertBlobPromise = Promise.resolve();
     if (buildSettings.uploadBinary && !imageBlob) {
-      imageBlob = blobConverters.dataUrlToBlob(shot.getClip(shot.clipNames()[0]).image.url);
-      shot.getClip(shot.clipNames()[0]).image.url = "";
+      let clipImage = shot.getClip(shot.clipNames()[0]).image
```
Comment (ianb) [inline addon/webextension/background/takeshot.js]: Nit: semicolon

File: addon/webextension/background/takeshot.js [hunk 2]
```diff
@@ -164,14 +192,19 @@ this.takeshot = (function() {
     });
   }
 
-  function uploadShot(shot, blob) {
+  function uploadShot(shot, blob, thumbnail) {
     let headers;
+    // @TODO: add thumbnail to payload
```
Comment (ianb) [inline addon/webextension/background/takeshot.js]: Is this still a todo?

File: server/src/pages/shot/controller.js [hunk 3]
```diff
@@ -120,32 +121,44 @@ exports.deleteShot = function(shot) {
...
+      method: 'POST',
+      credentials: 'include',
+      headers: new Headers({
+        'content-type': 'application/json'
+      }),
+      body
+    });
+    return fetch(req).then((resp) => {
+      if (!resp.ok) {
+        var errorMessage = "Error saving edited shot";
```
Comment (ianb) [inline server/src/pages/shot/controller.js]: Nit: `let` (and above)

File: server/src/server.js [hunk 4]
```diff
@@ -623,7 +623,9 @@ app.post("/api/set-login-cookie", function(req, res) {
   });
 });
 
-app.put("/data/:id/:domain", upload.single('blob'), function(req, res) {
+app.put("/data/:id/:domain",
+  upload.fields([{name: "blob", maxCount: 1}, {name: "thumbnail", maxCount: 1}]),
+  function(req, res) {
```
Comment (ianb) [inline server/src/server.js]: I guess the indentation below should be changed

File: shared/shot.js [hunk 5]
```diff
@@ -603,18 +603,19 @@ class AbstractShot {
...
 siteName openGraph twitterCard documentSize
-fullScreenThumbnail abTests
+thumbnail abTests
 `).split(/\s+/g);
 
 // Attributes that will be accepted in the constructor, but ignored/dropped
 AbstractShot.prototype.DEPRECATED_ATTRS = (`
 microdata history ogTitle createdDevice head body htmlAttrs bodyAttrs headAttrs
 readable hashtags comments showPage isPublic resources deviceId url
+fullScreenThumbnail
```
Comment (ianb) [inline shared/shot.js]: I didn't realize we had never deprecated fullScreenThumbnail (which was from when we were doing more expansive data capturing). Anyway, good to have it gone.

**Comment by chenba** [comment]
Yep, that TODO was a leftover I missed.  Squashed and merged.

**Pull request merged**